### PR TITLE
Create new instant answer for combinations and permutations

### PR DESCRIPTION
**Pull request by richardscollin**
**What does your Instant Answer do?**
Computes combinations and permutations.
10 choose 3 is 120
18 permute 2 is 306

**What problem does your Instant Answer solve (Why is it better than organic links)?**

**What is the data source for your Instant Answer? (Provide a link if possible)**
N/A

**Why did you choose this data source?**
N/A

**Are there any other alternative (better) data sources?**
N/A

**What are some example queries that trigger this Instant Answer?**
10 choose 2
18 permute 2
5 permutation 2

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
students, math

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
No.

**Which existing Instant Answers will this one supercede/overlap with?**

**Are you having any problems? Do you need our help with anything?**
Had an issue with getting it to work properly with european formated numbers.
There still may be a couple of issues with this.

**Where did you hear about DuckDuckHack? (For first time contributors)**
Can't remember. Hacker News?

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
screenshot
## Checklist

Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
[X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 7
        [] IE 8

    - Mac OSX
        [] Google Chrome
        [] Firefox
        [] Opera
        [] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```

### CODE REVIEW

File: lib/DDG/Goodie/Combination.pm [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+name "Combination";
+description "Computes combinations and permutations.";
+primary_example_queries "10 choose 3", "25 permute 16";
+secondary_example_queries "16 permutation 3";
+category "calculations";
+topics "math";
+code_url "https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Combination.pm";
+attribution github => ["richardscollin", "Collin Richards"];
+
+triggers any => "any", "choose", "permute", "permutation";
```
Comment (mintsoft) [inline lib/DDG/Goodie/Combination.pm]: How about adding nPr and nCr?

File: lib/DDG/Goodie/Combination.pm [hunk 2]
```diff
@@ -0,0 +1,90 @@
+package DDG::Goodie::Combination;
+# ABSTRACT: Compute combinations and permutations
+# Start at https://duck.co/duckduckhack/goodie_overview if you are new
```
Comment (mintsoft) [inline lib/DDG/Goodie/Combination.pm]: Can you bomb this comment except for the `# ABSTRACT` line above?

File: lib/DDG/Goodie/Combination.pm [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+attribution github => ["richardscollin", "Collin Richards"];
+
+triggers any => "any", "choose", "permute", "permutation";
+
+my $number_re = number_style_regex();
+
+# Handle statement
+handle query => sub {
+    my $query = $_;
+    return unless /^($number_re) (choose|permute|permutation) ($number_re)$/i;
```
Comment (mintsoft) [inline lib/DDG/Goodie/Combination.pm]: This would need updating too

File: lib/DDG/Goodie/Combination.pm [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    my $style = number_style_for($1,$3);
+    return unless $style; #Cannot determine number format
+    my $operation = lc $2;
+
+    if ($operation eq 'permutation') {
+        $operation = 'permute';#standardizes output for tests
+    }
+
+    my $p1 = $style->for_computation($1);
```
Comment (mintsoft) [inline lib/DDG/Goodie/Combination.pm]: Can these variable names be more explicit?

File: lib/DDG/Goodie/Combination.pm [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    my $p2 = $style->for_computation($3);
+
+    #Ensure both are non-negative integers
+    return unless $p1 =~ /^\d+$/ && $p2 =~ /^\d+$/;
+
+	#Do not try to calculate undefined combinations
```
Comment (mintsoft) [inline lib/DDG/Goodie/Combination.pm]: Can you harmonise the indentation here; it should be soft-tabbed, 4 spaces per "tab"

File: lib/DDG/Goodie/Combination.pm [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        $result = $style->for_display(choose($p1, $p2));
+	} else { #must be permute
+        $result = $style->for_display(permute($p1, $p2));
+	}
+
+    return $result,
+      structured_answer => {
+        input     => [$style->for_display($p1) . " $operation " . $style->for_display($p2)],
+        operation => $operation,
+        result    => $style->with_html($result),
```
Comment (mintsoft) [inline lib/DDG/Goodie/Combination.pm]: I think at the moment this will have to be without `with_html` as the output will get html_enc'd automatically

**Comment by mintsoft** [comment]
Hiya @richardscollin This looks awesome; I've put some first-pass feedback against the diff. Feel free to @ me if you have any questions!

**Comment by richardscollin** [comment]
Thanks @mintsoft ! Would you prefer me to do those changes in seperate commits or one large one? Which variables should be more explicit? Also I believe with_html is required because it adds html for superscripted exponents.

**Comment by mintsoft** [comment]
@richardscollin Not fussed; you can just commit more changes into your fork as you wish.

Variable wise; it's `$p1` and `$p2` I'm thinking about. Which is the larger set and which is the number chosen for example.

You're completely correct about the ideal being with_html; however the structured answers will html encode all output at the moment to protect against XSS. There should be some work coming soon to lift this limitation, however at the moment we're stuck!

File: t/Combination.t [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+use DDG::Test::Goodie;
+
+zci answer_type => "combination";
+zci is_cached   => 1;
+
+ddg_goodie_test(
+    [qw( DDG::Goodie::Combination )],
+    # At a minimum, be sure to include tests for all:
+    # - primary_example_queries
+    # - secondary_example_queries
```
Comment (MrChrisW) [inline t/Combination.t]: @richardscollin These comments can be removed. This IA looks awesome btw :+1:

File: lib/DDG/Goodie/Combination.pm [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    my $result;
+
+    if ('choose' eq $operation) {
+        $result = choose($n, $k);
+    } else { #must be permute
+        $result = permute($n, $k);
+    }
+
+    #Return no result if overflow
+    return if $result eq '-nan' or $result eq 'nan';
```
Comment (jagtalon) [inline lib/DDG/Goodie/Combination.pm]: Nice

**Comment by jagtalon** [comment]
Sorry for being late on this--we're checking it out!

**Comment by jagtalon** [comment]
@abeyang https://ddh1.duckduckgo.com/?q=10+choose+3&ia=answer what do you think? Looks pretty straightforward and uses our template.

**Comment by moollaza** [comment]
@richardscollin @jagtalon @mintsoft my only suggestion is that the `operation` should probably be, `calculate`, rather that `choose`, so that it reads "Calculate: 10 choose 3".

Otherwise LGTM :+1:

**Comment by jagtalon** [comment]
@richardscollin nice! :+1:

**Comment by moollaza** [comment]
@richardscollin congrats! It's live: https://duckduckgo.com/?q=10+choose+2&ia=answer

Thanks a lot for contributing, we really appreciate it :smiley:. 

Now that you've made your first Goodie, we really hope you'll continue to contribute and make more Instant Answers! Please checkout our [Ideas forum](https://duck.co/ideas) if you're unsure about what to build. There's also lots of Github Issues in our repos that need someone to tackle them!

**Pull request merged**